overlord/state/taskrunner.go

```diff
@@ -142,14 +147,24 @@ func (r *TaskRunner) Ensure() {
 	}
 }
 
-// Stop stops all concurrent activities and returns after that's done.
-// Note that Stop will lock the state.
+// Stop kills all concurrent activities and returns after that's done.
 func (r *TaskRunner) Stop() {
-	r.state.Lock()
-	defer r.state.Unlock()
+	r.mu.Lock()
+	defer r.mu.Unlock()
 
-	for _, tb := range r.tombs {
+	for id, tb := range r.tombs {
 		tb.Kill(nil)
 		tb.Wait()
+		delete(r.tombs, id)
+	}
+}
+
+// Wait waits for all concurrent activities and returns after that's done.
+func (r *TaskRunner) Wait() {
```

Comment:
Per conversation last night, this method cannot hold the lock while waiting.

Something like this should work:

s.mu.Lock()
defer s.mu.Unlock()
for len(r.tombs) > 0 {
        for id, t := range r.tombs {
                s.mu.Unlock()
                t.Wait()
                s.mu.Lock()
                delete(r.tombs, id)
                break
        }
}